frontend/CODE_GUIDELINES.md

```diff
@@ -0,0 +1,105 @@
+  - Focus on testing functionality.
+  - Simple components need tests too, mostly to make sure we don't break their imports.
+  - Testing is as much about easy refactors as anything. We need to be able to refactor confidently.
+- Don't be afraid to instrument your code for testing purposes.
+- Get good at mocking imports.
+  - Declare your mocks variables at the top of the file with the `let mock*` pattern.
+  - `jest.mock()` calls should follow the `let mock*` declarations.
+  - Jest automatically hoists your `jest.mock()` calls to the top of the file, so avoid confusion by keeping them near the top of the file. 
+  - Avoid `mockFn.mockImplementation` and `mockFn.mockImplementationOnce`. They're not **bad**, but they're more verbose and a bit harder to read than the `let mock*` pattern.
+
+## Javascript
+
+- Prefer `const` over `let`. Avoid `var`.
+- Prefer named functions, i.e., `function useSomething(){}` over anonymous functions.
+- Arrow functions are best for one-liners.
+- Write functional JS. `class` and `this` are unnecessary in most modern JavaScript.
+- Avoid mutating objects. This is bug-avoidance-rule-number-one in JavaScript.
+  - Object mutation is fine if it's visually close to the object declaration.
+  - There are times when object mutation simplifies the code, but BE CAREFUL AND EXPLICIT.
+- Be wary of using closure-scope for variable access. It's not functional and couples functions together, making refactors difficult.
+- Prefer "dot-syntax" over excessive destructuring.
+  - Well-crafted destructures can be very readable.
+  - Too much destructuring can kill your readability.
+  - For instance, always use `import constants from '~/constants';` This enables confident find/replace changes to the constants, because you can search for `constants.SOME_KEY`.
+  - Destructuring removes the context of the variable. The reader no longer knows where it came from. This is fine in vertically-adjacent code... but it can be hard to read otherwise.
+
+## Functions
+
+- Prefer a single return. There are definitely exceptions, but early returns can make debugging difficult.
+- Try to write small, single-purpose functions. Break functions down as much as is reasonable.
+- Prefer single-argument functions. 
+  - If multiple inputs are necessary, pass them in as a single object that can be destructured in the function signature.
+  - This is just a guideline, but fewer arguments is better.
+- Avoid nested conditionals when reasonable.
+
+## Comments
+
+- Comments lie and should be used sparingly.
+- If you're changing commented code, make sure to review and edit the comments carefully.
+- If you're about to commit a comment, see if you can avoid it with the following strategies:
+  - Replace comments with well-named, single-purpose functions where possible.
+  - Break multiple, ambiguous operations out into separate variable declarations.
+  - Prefer single variables in `if` and `switch` statements over inline operations.
+
+## React
```

Comment:
I also like trying to limit the size of react files to help with readability. Maybe one react file should only be exporting one Component?